vsintegration/src/FSharp.Editor/Navigation/GoToDefinitionService.fs

```diff
@@ -69,30 +70,91 @@ type internal FSharpGoToDefinitionService
-
-                                let text = MetadataAsSource.decompileCSharp(symbolFullTypeName, targetSymbolAssemblyFileName)
-                                let tmpShownDocOpt = metadataAsSourceService.ShowCSharpDocument(tmpProjInfo, tmpDocInfo, text)
-                                match tmpShownDocOpt with
-                                | Some tmpShownDoc ->
-                                    let navItem = FSharpGoToDefinitionNavigableItem(tmpShownDoc, TextSpan())                               
-                                    gtd.NavigateToItem(navItem, statusBar)
-                                    true
-                                | _ ->
-                                    statusBar.TempMessage (SR.CannotDetermineSymbol())
-                                    false
-                            with
+                    | FSharpGoToDefinitionResult.ExternalAssembly(targetSymbolUse, _targetExternalSymbol, metadataReferences) ->
+                        let textOpt =
+                            match targetSymbolUse.Symbol with
+                            | :? FSharpEntity as symbol ->
+                                symbol.TryGetMetadataText()
+                                |> Option.map (fun text -> text, symbol.DisplayName)
+                            | :? FSharpMemberOrFunctionOrValue as symbol ->
+                                symbol.ApparentEnclosingEntity.TryGetMetadataText()
+                                |> Option.map (fun text -> text, symbol.ApparentEnclosingEntity.DisplayName)
+                            | _ ->
+                                None
+
+                        match textOpt with
+                        | Some (text, fileName) ->
+                            let tmpProjInfo, tmpDocInfo = 
+                                MetadataAsSource.generateTemporaryDocument(
+                                    AssemblyIdentity(targetSymbolUse.Symbol.Assembly.QualifiedName), 
+                                    fileName, 
+                                    metadataReferences)
+                            let tmpShownDocOpt = metadataAsSourceService.ShowDocument(tmpProjInfo, tmpDocInfo.FilePath, SourceText.From(text.ToString()))
+                            match tmpShownDocOpt with
+                            | Some tmpShownDoc ->
+                                let goToAsync =
+                                    asyncMaybe {
+                                        let userOpName = "TryGoToDefinition"
+                                        let! _, _, projectOptions = projectInfoManager.TryGetOptionsForDocumentOrProject (tmpShownDoc, cancellationToken, userOpName)
+                                        let! _, _, checkResults = checkerProvider.Checker.ParseAndCheckDocument(tmpShownDoc, projectOptions, userOpName)
+                                        let! r =
+                                            let rec areTypesEqual (ty1: FSharpType) (ty2: FSharpType) =
```

Comment:
Agree with the commment below. We should file an issue to figure out a way to lift this into FCS proper.

Reply:
Equality does exist for symbols, but the symbols from the generated metadata may be malformed, or just not exactly match the requirements for the existing equality in FCS. Basically, we need `NicePrint` to generate signatures that can be equatable.